components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/token/ExtendedJWTBearerGrantHandler.java

```diff
@@ -0,0 +1,199 @@
+ *
+ *WSO2 Inc. licenses this file to you under the Apache License,
+ *Version 2.0 (the "License"); you may not use this file except
+ *in compliance with the License.
+ *You may obtain a copy of the License at
+ *
+ *http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *Unless required by applicable law or agreed to in writing,
+ *software distributed under the License is distributed on an
+ *"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *KIND, either express or implied.  See the License for the
+ *specific language governing permissions and limitations
+ *under the License.
+ */
+
+package org.wso2.carbon.apimgt.keymgt.token;
+
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.apimgt.keymgt.ScopesIssuer;
+import org.wso2.carbon.apimgt.keymgt.handlers.ResourceConstants;
+import org.wso2.carbon.identity.application.authentication.framework.model.AuthenticatedUser;
+import org.wso2.carbon.identity.application.authentication.framework.util.FrameworkUtils;
+import org.wso2.carbon.identity.application.common.model.ClaimMapping;
+import org.wso2.carbon.identity.application.common.model.IdentityProvider;
+import org.wso2.carbon.identity.application.common.model.PermissionsAndRoleConfig;
+import org.wso2.carbon.identity.application.common.model.RoleMapping;
+import org.wso2.carbon.identity.application.common.util.IdentityApplicationConstants;
+import org.wso2.carbon.identity.oauth.config.OAuthServerConfiguration;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.grant.jwt.JWTBearerGrantHandler;
+import org.wso2.carbon.identity.oauth2.grant.jwt.JWTConstants;
+import org.wso2.carbon.identity.oauth2.model.RequestParameter;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+import org.wso2.carbon.idp.mgt.IdentityProviderManagementException;
+import org.wso2.carbon.idp.mgt.IdentityProviderManager;
+
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Extends the JWTBearerGrantHanlder, in which the validateScope() method is overridden
+ * to implement role based scope validation
+ */
+public class ExtendedJWTBearerGrantHandler extends JWTBearerGrantHandler {
+
+    private static Log log = LogFactory.getLog(ExtendedJWTBearerGrantHandler.class);
+    private IdentityProvider identityProvider = null;
+
+    @Override public boolean validateScope(OAuthTokenReqMessageContext tokReqMsgCtx) {
+
+        SignedJWT signedJWT = null;
+        JWTClaimsSet claimsSet = null;
+        String[] roles = null;
+        try {
+            signedJWT = getSignedJWT(tokReqMsgCtx);
+        } catch (IdentityOAuth2Exception e) {
+            log.error("Couldn't retrieve signed JWT", e);
```

Comment:
Why don't we throw an exception here? Now if an error occurred and signedJWT becomes null, then in next step >> claimsSet = getClaimSet(signedJWT); // This will cause null pointer errors